src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala

```diff
@@ -170,71 +177,88 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
+          val inAssetsTry = ErgoTransaction.extractAssets(boxesToSpend)
+          inAssetsTry match {
+            case Success((inAssets, inAssetsNum)) =>
+              lazy val newAssetId = ByteArrayWrapper(inputs.head.boxId)
+              val tokenAccessCost = stateContext.currentParameters.tokenAccessCost
+              val currentTxCost = validation.result.payload.get
+              // Cost of assets preservation rules checks.
+              // We iterate through all assets to create a map (cost: `(outAssetsNum + inAssetsNum) * tokenAccessCost)`)
+              // and after that we iterate through unique asset ids to check preservation rules (cost: `(inAssets.size + outAssets.size) * tokenAccessCost`)
+              val totalAssetsAccessCost = (outAssetsNum + inAssetsNum) * tokenAccessCost +
+                (inAssets.size + outAssets.size) * tokenAccessCost
+              val newCost = addExact(currentTxCost, totalAssetsAccessCost)
+
+              validation
+                // Check that transaction is not too costly considering all the assets
+                .validate(bsBlockTransactionsCost, maxCost >= newCost, s"$id: assets cost")
+                .validateSeq(outAssets) {
+                  case (validationState, (outAssetId, outAmount)) =>
+                    val inAmount: Long = inAssets.getOrElse(outAssetId, -1L)
+
+                    // Check that for each asset output amount is no more than input amount,
+                    // with a possible exception for a new asset created by the transaction
+                    validationState.validate(txAssetsPreservation,
+                      inAmount >= outAmount || (outAssetId == newAssetId && outAmount > 0),
+                      s"$id: Amount in = $inAmount, out = $outAmount. Allowed new asset = $newAssetId, out = $outAssetId")
+                }
+                .payload(newCost)
+            case Failure(e) =>
+              // should never be here as far as we've already checked this when we've created the box
+              ModifierValidator.fatal(e.getMessage)
+          }
       }
       // Check inputs, the most expensive check usually, so done last.
       .validateSeq(boxesToSpend.zipWithIndex) { case (validation, (box, idx)) =>
         val currentTxCost = validation.result.payload.get
 
-        val input = inputs(idx)
-        val proof = input.spendingProof
-        val transactionContext = TransactionContext(boxesToSpend, dataBoxes, this)
-        val inputContext = InputContext(idx.toShort, proof.extension)
-
-        val ctx = new ErgoContext(
-          stateContext, transactionContext, inputContext,
-          costLimit = maxCost - currentTxCost, // remaining cost so far
-          initCost = 0)
-
-        val costTry = verifier.verify(box.ergoTree, ctx, proof, messageToSign)
-        costTry.recover { case t =>
-          log.debug(s"Tx verification failed: ${t.getMessage}")
+        val (res, time) = measureTimeNano {
```

Comment:
Time is measured even if metrics collection is off (which will the case for 99+% of the nodes). I guess this measurement should be optional.

Reply:
fixed